private/buf/buffetch/internal/reader.go

```diff
@@ -583,101 +701,190 @@ func findTerminateFileDirectoryPathFromBucket(
-func findTerminateFileDirectoryPathFromOS(subDirPath string, terminateFileNames []string) (string, error) {
+// the root of the local filesystem.
+func findTerminateFileDirectoryPathFromOS(
+	subDirPath string,
+	terminateFileNames [][]string,
+) (TerminateFileProvider, error) {
 	if len(terminateFileNames) == 0 {
-		return "", nil
+		return nil, nil
 	}
 	fileInfo, err := os.Stat(normalpath.Unnormalize(subDirPath))
 	if err != nil {
 		if os.IsNotExist(err) {
-			return "", storage.NewErrNotExist(subDirPath)
+			return nil, storage.NewErrNotExist(subDirPath)
 		}
-		return "", err
+		return nil, err
 	}
 	if !fileInfo.IsDir() {
-		return "", normalpath.NewError(normalpath.Unnormalize(subDirPath), errors.New("not a directory"))
+		return nil, normalpath.NewError(normalpath.Unnormalize(subDirPath), errors.New("not a directory"))
 	}
+	terminateFiles := make([]TerminateFile, len(terminateFileNames))
 	terminateFileDirectoryPath, err := normalpath.NormalizeAndAbsolute(subDirPath)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
+	foundFiles := 0
 	for {
-		fullTerminateFileNames := make([]string, len(terminateFileNames))
-		for i, terminateFileName := range terminateFileNames {
-			fullTerminateFileNames[i] = normalpath.Unnormalize(normalpath.Join(terminateFileDirectoryPath, terminateFileName))
-		}
-		exists, err := anyExistsOS(fullTerminateFileNames)
+		foundTerminateFiles, err := terminateFilesOnOS(terminateFileDirectoryPath, terminateFileNames)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		if exists {
-			return terminateFileDirectoryPath, nil
+		// More terminate files were found than layers of hierarchy, return an error. This path
+		// should be unreachable.
+		if len(foundTerminateFiles) > len(terminateFileNames) {
+			return nil, fmt.Errorf("more than one terminate file found per level of prioritization: %v", foundTerminateFiles)
+		}
+		for i, terminateFile := range foundTerminateFiles {
+			// We only want to return the first file for a hiearchy, so if a file already exists
+			// for a layer of hierarchy, we do not check again.
+			if terminateFiles[i] == nil {
+				// If a file is found for a specific layer of hierarchy, we return the first file found
+				// for that hierarchy.
+				if terminateFile != nil {
+					terminateFiles[i] = terminateFile
+					foundFiles++
+				}
+			}
+		}
+		// If we have found a terminate file for each layer, we can return now.
+		if foundFiles == len(terminateFileNames) {
+			return newTerminateFileProvider(terminateFiles), nil
 		}
 		parent := normalpath.Dir(terminateFileDirectoryPath)
 		if parent == terminateFileDirectoryPath {
-			return "", nil
+			break
 		}
 		terminateFileDirectoryPath = parent
 	}
+	// The number of terminate files found is less than the number of layers of terminate files
+	// we are accepting, so we must prune the nil values from the initial instantiation of the slice.
+	prunedTerminateFilesList := []TerminateFile{}
```

Comment:
For default slice instantiation, prefer `var prunedTerminateFiles []TerminateFile`.